tv_apps/smart-home/js/home.js

```diff
@@ -130,9 +157,71 @@
 
-    onCardInserted: function(card, idx) {
-      this.cardScrollable.insertNodeBefore(this._createCardNode(card), idx);
+    _onPinInit: function(newCardButtonElem) {
+      newCardButtonElem.classList.add('new-card');
+      newCardButtonElem.classList.add('new-card-transition');
+
+      // add sliding transition to card-list and every existing cards
+      var i;
+      var cardElem;
+      for(i = 0; i < this.cardListElem.children.length; i++) {
+        cardElem = this.cardListElem.children[i];
+        cardElem.classList.add('card-slide');
+      }
+      this.cardListElem.classList.add('card-slide');
+    },
+
+    _onCardListSlideEnd: function(newCardButtonElem, idx) {
+      this.cardScrollable.setOtherNodesPosition(idx);
       this.cardScrollable.focus(idx);
+      this.cardListElem.classList.remove('card-slide');
+      newCardButtonElem.classList.remove('new-card');
+    },
+
+    onCardInserted: function(card, idx) {
+      var newCardElem = this._createCardNode(card);
+      var newCardButtonElem = newCardElem.children[0];
+      this.cardScrollable.spatialNavigator.getFocusedElement().blur();
+      this._onPinInit(newCardButtonElem);
+
+      // insert new card into cardScrollable
+      var newItem = this.cardScrollable.insertNodeBefore(newCardElem, idx);
+      var originalDuration = getComputedStyle(this.cardListElem)
+                             .transitionDuration;
+
+      // all the transition end states are handled by card-list
+      var transitionHandler = function(evt) {
+        if (evt.propertyName === 'transform') {
+          switch(evt.target) {
+            case this.cardListElem:
+              this._onCardListSlideEnd(newCardButtonElem, idx);
+              break;
+            case newCardButtonElem:
+              // end of pin-card transition, remove all the classes we added
+              var i;
+              for(i = 0; i < this.cardListElem.children.length - 1; i++) {
+                this.cardListElem.children[i].classList.remove('card-slide');
+              }
+              newCardButtonElem.classList.remove('new-card-transition');
+              this.cardListElem.style.transitionDuration = originalDuration;
+              this.cardListElem.removeEventListener(
+                                   'transitionend', transitionHandler);
+              break;
+          }
+        }
+      }.bind(this);
+      this.cardListElem.addEventListener('transitionend', transitionHandler);
+
+      // start scrolling card list.
+      // Also consider the case when the card-list does not need to scroll
+      var prevTransform = this.cardListElem.style.transform;
+      var distance = Math.abs(this.cardScrollable.getScrollDistance(newItem));
+      this.cardListElem.style.transitionDuration = distance / 1000 + 's';
+      this.cardScrollable.scrollTo(newItem);
+      if (!prevTransform ||
+          prevTransform === this.cardListElem.style.transform) {
+        this._onCardListSlideEnd(newCardButtonElem, idx);
```

Comment:
Please consider to put codes that accesses `cardListElem.style` to somewhere inside `Scrollable` or another independent module that acts like a helper of `Scrollable`.